presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -328,23 +323,51 @@ private <T extends Node> T processSameType(T node)
 
     private QuerySpecification rewriteQuerySpecificationIfCompatible(QuerySpecification querySpecification, Table baseTable)
     {
-        QualifiedObjectName tableName = createQualifiedObjectName(session, baseTable, baseTable.getName());
-        List<ConnectorMaterializedViewDefinition> linkedMaterializedViews = baseTableToMaterializedViews.getOrDefault(tableName, ImmutableList.of());
+        Optional<String> errorMessage = MaterializedViewRewriteQueryShapeValidator.validate(querySpecification);
+
+        if (errorMessage.isPresent()) {
+            log.info(errorMessage.get());
+            return querySpecification;
+        }
+
+        List<QualifiedObjectName> referencedMaterializedViews = metadata.getReferencedMaterializedViews(
+                session,
+                createQualifiedObjectName(session, baseTable, baseTable.getName()));
 
-        for (ConnectorMaterializedViewDefinition materializedViewDefinition : linkedMaterializedViews) {
-            Table materializedViewTable = new Table(QualifiedName.of(materializedViewDefinition.getTable()));
-            Query materializedViewQuery = (Query) sqlParser.createStatement(materializedViewDefinition.getOriginalSql(), createParsingOptions(session));
+        // TODO: Select the most compatible and efficient materialized view for query rewrite optimization https://github.com/prestodb/presto/issues/16431
+        // TODO: Refactor query optimization code https://github.com/prestodb/presto/issues/16759
```

Comment:
Most of this TODO should have been resolved with this PR? If yes can we update the github issue

Reply:
Was planning on updating said issue once the PR is merged, but I don't think it's 100% resolved.